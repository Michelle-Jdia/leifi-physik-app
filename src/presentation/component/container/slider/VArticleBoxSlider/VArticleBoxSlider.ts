import { link } from '@/presentation/static/link';

export const articleBoxSliderProps = {
    lastBox: {
        title: 'Zu deinen Favoriten',
        link,
    },
};
