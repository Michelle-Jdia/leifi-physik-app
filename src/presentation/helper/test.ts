import type { ComponentMountingOptions } from '@vue/test-utils';
import type { Component } from 'vue';
import { config, mount } from '@vue/test-utils';
import { expect, test } from 'vitest';

beforeAll(() => {
    config.global.renderStubDefaultSlot = true;
});

afterAll(() => {
    config.global.renderStubDefaultSlot = false;
});

export const testSnapshot = (
    component: Component,
    options?: ComponentMountingOptions<object>,
): void => {
    test('snapshot', () => {
        const wrapper = mount(component, {
            shallow: true,
            ...options,
        });

        expect(wrapper.element).toMatchSnapshot();
    });
};
