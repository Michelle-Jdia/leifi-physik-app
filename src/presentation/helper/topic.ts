import type { Topic } from '@/data/type/app/topic';
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import { RouteName } from '@/connection/router/routeName';

export function createTopicSlider(topic: Topic): SliderItem {
    return {
        img: {
            src: topic.image?.url || '',
            alt: topic.image?.alt || '',
        },
        title: topic.name,
        link: {
            name: RouteName.TOPIC_ID,
            params: {
                id: topic.id,
            },
        },
    };
}
