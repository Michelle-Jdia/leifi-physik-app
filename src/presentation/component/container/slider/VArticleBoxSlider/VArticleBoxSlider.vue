<script setup lang="ts">
import type { FavoriteTypeKey } from '@/data/type/app/favorite';
import type { ClassValue } from 'clsx';
import type { RouteLocationNamedRaw } from 'vue-router';
import { SwiperSlide } from 'swiper/vue';
// import type { Props as VArticleBoxProps } from '@/molecule/box/VArticleBox/VArticleBox.vue';
import { cn } from '@/presentation/helper/style';
import VFavoriteWrapper from '@/container/wrapper/VFavoriteWrapper/VFavoriteWrapper.vue';
import VArticleBox from '@/molecule/box/VArticleBox/VArticleBox.vue';
import VBadge from '@/atom/badge/VBadge/VBadge.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VSlider from '@/atom/slider/VSlider/VSlider.vue';

export interface SliderItem {
    id: string;
    type: FavoriteTypeKey;
    // isLearned?: VArticleBoxProps['isLearned'];
    // withNotification?: VArticleBoxProps['withNotification'];
}

interface Props {
    lastBox: {
        title: string;
        link: Partial<RouteLocationNamedRaw>;
    };
    sliderItems: SliderItem[];
    modifier?: ClassValue;
}

defineProps<Props>();
</script>

<template>
    <div :class="cn('article-box-slider', modifier)">
        <h2 v-if="$slots.default" class="v-h2 v-mb-box md:v-mb-box-xl">
            <slot />
        </h2>

        <v-slider modifier="!v-overflow-visible">
            <swiper-slide v-for="item in sliderItems" :key="`${item.id}`" class="!v-h-auto">
                <v-favorite-wrapper :id="item.id" :type="item.type">
                    <template #default="{ title, subtitle, branch, link, isFavorite }">
                        <router-link :to="link" class="v-h-full v-w-full">
                            <v-article-box
                                :title="title"
                                :subtitle="subtitle"
                                :is-favorite="isFavorite"
                                :is-learned="true"
                                :with-notification="false"
                                modifier="v-h-full v-w-full"
                            >
                                <v-badge :modifier="branch.color">
                                    {{ branch.title }}
                                </v-badge>
                            </v-article-box>
                        </router-link>
                    </template>
                </v-favorite-wrapper>
            </swiper-slide>

            <swiper-slide class="!v-h-auto">
                <router-link :to="lastBox.link" class="v-h-full v-w-full">
                    <v-article-box
                        modifier="v-bg-transparent v-shadow-none v-border-dashed v-h-full v-w-full"
                        :is-learned="true"
                        :title="lastBox.title"
                    >
                        <v-icon src="icon/arrow-right.svg" size="sm" />
                    </v-article-box>
                </router-link>
            </swiper-slide>
        </v-slider>
    </div>
</template>
