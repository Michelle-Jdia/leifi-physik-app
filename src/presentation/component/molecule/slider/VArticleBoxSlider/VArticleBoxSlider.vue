<script setup lang="ts">
import type { ClassValue } from 'clsx';
import type { RouteLocationRaw } from 'vue-router';
import type { Props as VArticleBoxProps } from '@/molecule/box/VArticleBox/VArticleBox.vue';
import { cn } from '@/presentation/helper/style';
import { SwiperSlide } from 'swiper/vue';
import VSlider from '@/atom/slider/VSlider/VSlider.vue';
import VBadge from '@/atom/badge/VBadge/VBadge.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VArticleBox from '@/molecule/box/VArticleBox/VArticleBox.vue';

export interface SliderItem extends VArticleBoxProps {
    badge: {
        title: string;
        color: string;
    };
    link: Partial<RouteLocationRaw>;
}

interface Props {
    lastBox: {
        title: string;
        link: Partial<RouteLocationRaw>;
    };
    sliderItems: SliderItem[];
    modifier?: ClassValue;
}

defineProps<Props>();
</script>

<template>
    <div :class="cn('article-box-slider v-p-box-xl', modifier)">
        <h2 v-if="$slots.default" class="v-h2 v-mb-box">
            <slot />
        </h2>

        <v-slider modifier="!v-overflow-visible">
            <swiper-slide
                v-for="(item, index) in sliderItems"
                :key="`${item.title}-${index}`"
                class="!v-h-auto"
            >
                <router-link :to="item.link" class="v-h-full v-w-full">
                    <v-article-box
                        :title="item.title"
                        :subtitle="item.subtitle"
                        :is-learned="item.isLearned"
                        :with-notification="item.withNotification"
                        modifier="v-h-full v-w-full"
                    >
                        <v-badge :modifier="item.badge.color">
                            {{ item.badge.title }}
                        </v-badge>
                    </v-article-box>
                </router-link>
            </swiper-slide>

            <swiper-slide>
                <router-link :to="lastBox.link" class="v-h-full v-w-full">
                    <v-article-box
                        modifier="v-border-dashed v-bg-transparent v-shadow-none"
                        is-learned
                        :title="lastBox.title"
                    >
                        <v-icon src="icon/arrow-right.svg" size="sm" />
                    </v-article-box>
                </router-link>
            </swiper-slide>
        </v-slider>
    </div>
</template>
