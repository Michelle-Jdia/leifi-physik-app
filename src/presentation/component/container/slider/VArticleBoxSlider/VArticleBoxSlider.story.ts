import type { Meta, StoryFn } from '@storybook/vue3';
import { ContentType } from '@/data/type/helper/contentType';
import { topic } from '@/presentation/static/topic';
import { articleBoxSliderProps } from '@/container/slider/VArticleBoxSlider/VArticleBoxSlider';
import VArticleBoxSlider from '@/container/slider/VArticleBoxSlider/VArticleBoxSlider.vue';

export default {
    component: VArticleBoxSlider,
    tags: ['autodocs'],
} as Meta<typeof VArticleBoxSlider>;

const Template: StoryFn<typeof VArticleBoxSlider> = (args) => ({
    components: {
        VArticleBoxSlider,
    },
    setup() {
        return {
            args,
        };
    },
    template: `
        <v-article-box-slider v-bind="args">
          ${topic.title}
        </v-article-box-slider>
    `,
});

export const Base = Template.bind({});

const ids = ['4958cb2c-00a1-465e-ab41-1cc048b7e90b', '77c55ea1-5d06-4c91-9f51-b3ccd24be1c7'];

Base.args = {
    lastBox: articleBoxSliderProps.lastBox,
    sliderItems: ids.map((id) => ({
        id,
        type: ContentType.ISSUE,
    })),
    modifier: 'bg-green-50',
};
