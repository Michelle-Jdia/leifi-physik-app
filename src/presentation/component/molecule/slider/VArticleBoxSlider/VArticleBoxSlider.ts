import { link } from '@/presentation/static/link';
import { topic } from '@/presentation/static/topic';
import { branch } from '@/presentation/static/branch';

export const articleBoxSliderProps = {
    lastBox: {
        title: 'Zu deinen Favoriten',
        link,
    },
    sliderItem: {
        title: topic.title,
        subtitle: topic.subtitle,
        link,
        badge: {
            title: branch.title,
        },
    },
};
