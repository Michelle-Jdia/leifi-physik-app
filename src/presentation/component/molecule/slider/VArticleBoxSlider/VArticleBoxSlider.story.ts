import type { Meta, StoryFn } from '@storybook/vue3';
import { articleBoxSliderProps } from '@/molecule/slider/VArticleBoxSlider/VArticleBoxSlider';
import { topic } from '@/presentation/static/topic';
import VArticleBoxSlider from '@/molecule/slider/VArticleBoxSlider/VArticleBoxSlider.vue';

export default {
    component: VArticleBoxSlider,
    tags: ['autodocs'],
} as Meta<typeof VArticleBoxSlider>;

const Template: StoryFn<typeof VArticleBoxSlider> = (args) => ({
    components: {
        VArticleBoxSlider,
    },
    setup() {
        return {
            args,
        };
    },
    template: `
        <v-article-box-slider v-bind="args">
          ${topic.title}
        </v-article-box-slider>
    `,
});

export const Base = Template.bind({});

const colors = ['bg-yellow-50', 'bg-pink-50', 'bg-blue-50'];

Base.args = {
    lastBox: articleBoxSliderProps.lastBox,
    sliderItems: colors.map((color, index) => ({
        title: articleBoxSliderProps.sliderItem.title,
        subtitle: articleBoxSliderProps.sliderItem.subtitle,
        link: articleBoxSliderProps.sliderItem.link,
        badge: {
            title: articleBoxSliderProps.sliderItem.badge.title,
            color,
        },
        isLearned: Boolean(index % 2),
        withNotification: Boolean(index % 3),
    })),
    modifier: 'bg-green-50',
};
