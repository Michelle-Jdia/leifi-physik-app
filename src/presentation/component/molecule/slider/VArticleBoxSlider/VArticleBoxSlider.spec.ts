import { render } from '@testing-library/vue';
import { testSnapshot } from '@/presentation/helper/test';
import { describe, expect, test } from 'vitest';
import { topic } from '@/presentation/static/topic';
import { articleBoxSliderProps } from '@/molecule/slider/VArticleBoxSlider/VArticleBoxSlider';
import VArticleBoxSlider from '@/molecule/slider/VArticleBoxSlider/VArticleBoxSlider.vue';

describe('VArticleBoxSlider', () => {
    const options = {
        slots: {
            default: topic.copy,
        },
        props: {
            lastBox: articleBoxSliderProps.lastBox,
            sliderItems: [
                articleBoxSliderProps.sliderItem,
                articleBoxSliderProps.sliderItem,
                articleBoxSliderProps.sliderItem,
            ],
        },
        global: {
            stubs: {
                ionIcon: true,
            },
        },
    };

    test('shows text that was passed to default slot', () => {
        const { getAllByText } = render(VArticleBoxSlider, options);

        getAllByText(topic.copy);

        expect(getAllByText(articleBoxSliderProps.sliderItem?.title as string)).toHaveLength(3);
    });

    test('doesnt render defaul slot if nothing was passed to it', () => {
        const { queryByText } = render(VArticleBoxSlider, {
            ...options,
            slots: {},
        });

        expect(queryByText(topic.copy)).toBeFalsy();
    });

    testSnapshot(VArticleBoxSlider, options);
});
