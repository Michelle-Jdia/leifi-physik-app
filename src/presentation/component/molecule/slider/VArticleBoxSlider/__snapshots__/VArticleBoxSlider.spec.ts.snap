// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VArticleBoxSlider > snapshot 1`] = `
<div
  class="article-box-slider v-p-box-xl"
>
  <h2
    class="v-h2 v-mb-box"
  >
    
    Geschwindigkeit bei gleichf√∂rm-iger Bewegung
    
  </h2>
  <v-slider-stub
    modifier="!v-overflow-visible"
    modules=""
    slidesperview="2.3"
    slidesperviewtablet="5.3"
    spacebetweenslides="8"
  >
    
    <swiper-slide-stub
      class="!v-h-auto"
      lazy="false"
      tag="div"
    >
      <router-link
        class="v-h-full v-w-full"
        to="[object Object]"
      >
        <v-article-box-stub
          modifier="v-h-full v-w-full"
          subtitle="Grundwissen"
          title="Masse und Volumen"
        >
          <v-badge-stub>
            Mechanik
          </v-badge-stub>
        </v-article-box-stub>
      </router-link>
    </swiper-slide-stub>
    <swiper-slide-stub
      class="!v-h-auto"
      lazy="false"
      tag="div"
    >
      <router-link
        class="v-h-full v-w-full"
        to="[object Object]"
      >
        <v-article-box-stub
          modifier="v-h-full v-w-full"
          subtitle="Grundwissen"
          title="Masse und Volumen"
        >
          <v-badge-stub>
            Mechanik
          </v-badge-stub>
        </v-article-box-stub>
      </router-link>
    </swiper-slide-stub>
    <swiper-slide-stub
      class="!v-h-auto"
      lazy="false"
      tag="div"
    >
      <router-link
        class="v-h-full v-w-full"
        to="[object Object]"
      >
        <v-article-box-stub
          modifier="v-h-full v-w-full"
          subtitle="Grundwissen"
          title="Masse und Volumen"
        >
          <v-badge-stub>
            Mechanik
          </v-badge-stub>
        </v-article-box-stub>
      </router-link>
    </swiper-slide-stub>
    
    <swiper-slide-stub
      lazy="false"
      tag="div"
    >
      <router-link
        class="v-h-full v-w-full"
        to="[object Object]"
      >
        <v-article-box-stub
          islearned="true"
          modifier="v-border-dashed v-bg-transparent v-shadow-none"
          title="Zu deinen Favoriten"
          withnotification="false"
        >
          <v-icon-stub
            modifier="false"
            size="sm"
            src="icon/arrow-right.svg"
            withspaceleft="false"
            withspaceright="false"
          />
        </v-article-box-stub>
      </router-link>
    </swiper-slide-stub>
  </v-slider-stub>
</div>
`;
